### ISSUE TITLE
[📍] Automatically apply fixes to failing ranges

### ISSUE THREAD

**Issue by dave-2** [catapult-project/catapult#4335]
From #3516.

## Problem
We often run into commit ranges where the test is busted. To continue bisecting, we want to automatically apply a fix the failing range. For many jobs, the failure covers only a part of the range, so it's less useful to apply the patch to the entire job. We also want this behavior to be visible in the UI, and to support multiple overlapping fixes (dependent on [support in bot_update](https://bugs.chromium.org/p/chromium/issues/detail?id=820653)).

## Suggested solution
* Currently, `Change.patch` refers to the patch that the user is currently trying. To distinguish fixes in the UI, we want them as a separate, new field in Change.
* Have an ndb Model where they keys are `git_hash` and the contents are a list of `Patch`es. Using the `git_hash` as the key allows us to look up fixes in constant time.
* Admins can add a fix using a webpage.
* Make `JobState.AddChange()` look up the Change's commits and add fixes to it.
* Show the fixes in the UI.

@perezju @simonhatch @anniesullie

**Comment by simonhatch** [comment]
That sounds good, can't wait to see this in action.

**Comment by perezju** [comment]
Sounds awesome!

That means to apply a fix to a range you would add it to each of the commits in the range? (That sounds fine too, just wanted to double check.)

**Comment by dave-2** [comment]
Yes, when an admin adds a new fix, it will look up the commit range and store the patch for each commit. So we're doing the range lookup at store time instead of fetch time.

**Comment by dave-2** [comment]
The drawback to the above is that it's costly to have patches cover a very large range. (And we can't really support it anyway because bot_update doesn't support multiple patches yet.)

**Comment by dave-2** [comment]
Here's a list of bugs with jobs to re-run after we have this implemented:

https://bugs.chromium.org/p/chromium/issues/detail?id=814183
https://bugs.chromium.org/p/chromium/issues/detail?id=814187
https://bugs.chromium.org/p/chromium/issues/detail?id=816587
https://bugs.chromium.org/p/chromium/issues/detail?id=816580
https://bugs.chromium.org/p/chromium/issues/detail?id=814199
https://bugs.chromium.org/p/chromium/issues/detail?id=814196
https://bugs.chromium.org/p/chromium/issues/detail?id=814195

**Comment by dave-2** [comment]
From Juan's comments on the CL and our followup discussion in Speed Services on 2/28:
https://chromium-review.googlesource.com/c/catapult/+/939664

* There may be weird edge cases where we add a new auto-fix while jobs are running, so those jobs will be half fixed and half not. We will have to add some kind of workaround for these.
* We can't support overlapping fixes for now, since our builders can't support multiple patches. We will have to reject overlapping fixes.
* We will need to be judicious in applying fixes. Generally, they should be fixes that cause an entire config's tests to break, and definitely not fixes that modify performance.
